packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts

```diff
@@ -0,0 +1,326 @@
+
+/**
+ * Utility enum for Schema class
+ */
+enum SchemaMode {
+  FILE = 'FILE',
+  CODE = 'CODE',
+  S3 = 'S3',
+};
+
+export class SchemaImpl implements Schema {
+
+  /**
+   * The definition for this schema
+   */
+  public readonly definition: string;
+
+  /**
+    * The underlying CFN GraphQL Schema Resource. This
+    * resource is created on `bind` which allows the Schema
+    * to be created outside the CDK (a feature that works well
+    * with the code-first approach).
+    *
+    * @internal
+    */
+  protected _schema!: CfnGraphQLSchema;
+
+  /**
+    * The SchemaMode utilized by the Schema Implementation
+    *
+    * @internal
+    */
+  protected _mode: SchemaMode;
+
+  public constructor() {
+    this.definition = '';
+    this._mode = SchemaMode.FILE;
+  }
+
+  /**
+    * Called when the GraphQL Api is initialized to allow this object to bind
+    * to the stack.
+    */
+  public bind(api: GraphqlApi): CfnGraphQLSchema {
+    if (!this._schema) {
+      this._schema = new CfnGraphQLSchema(api, 'Schema', {
+        apiId: api.apiId,
+        definition: this.definition,
+      });
+    }
+    return this._schema;
+  }
+
+  /**
+    * Escape hatch to add to Schema as desired. Will always result
+    * in a newline.
+    */
+  public addToSchema(_addition: string, _delimiter?: string): void {
+    throw new Error('API cannot append to schema because schema definition mode is not configured as CODE.');
+  }
+
+  /**
+    * Add a query field to the schema's Query. CDK will create an
+    * Object Type called 'Query'. For example,
+    *
+    * type Query {
+    *   fieldName: Field.returnType
+    * }
+    */
+  public addQuery(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add query. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+    * Add a mutation field to the schema's Mutation. CDK will create an
+    * Object Type called 'Mutation'. For example,
+    *
+    * type Mutation {
+    *   fieldName: Field.returnType
+    * }
+    */
+  public addMutation(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add mutation. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+    * Add a subscription field to the schema's Subscription. CDK will create an
+    * Object Type called 'Subscription'. For example,
+    *
+    * type Subscription {
+    *   fieldName: Field.returnType
+    * }
+    */
+  public addSubscription(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add subscription. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+    * Add type to the schema
+    */
+  public addType(_type: IIntermediateType): IIntermediateType {
+    throw new Error('API cannot add type because schema definition mode is not configured as CODE.');
```

Comment:
wow.. i honestly did not know you could do this 😱  my mind is actually blown what

Reply:
question, how would a user add this to the GraphqlApi? Because the Graphql API has the option of adding queries, and mutations to the schema as well!

https://github.com/aws/aws-cdk/blob/6b639bee925a9e552c405fde9d221d515e6023c3/packages/%40aws-cdk/aws-appsync/lib/graphqlapi.ts#L577-L643

If i set the type for schema to be a `Schema` it is immutable and thus cannot be targeted by the API. I guess what we **could** do is remove that feature entirely? Prevents new users from tripping and trying to add a Query if they are importing their schema locally.

Reply:
Maybe keep the API for now, but move the error code to `GraphqlApi` (basically, each of the methods does an `instanceof` check, and throws if the `schema` is not a `MutableSchema`)?